man/systemd-nspawn.xml

```diff
@@ -734,6 +734,31 @@
         capabilities are passed using the <command>--capabilities=</command>.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--rlimit=</option></term>
+
+        <listitem><para>Sets the specified POSIX resource limit for the container payload. Expects an assignment of the
+        form
+        <literal><replaceable>LIMIT</replaceable>=<replaceable>SOFT</replaceable>:<replaceable>HARD</replaceable></literal>
+        or <literal><replaceable>LIMIT</replaceable>=<replaceable>VALUE</replaceable></literal>, where
+        <replaceable>LIMIT</replaceable> should refer to a resource limit type, such as
+        <constant>RLIMIT_NOFILE</constant> or <constant>RLIMIT_NICE</constant>. The <replaceable>SOFT</replaceable> and
+        <replaceable>HARD</replaceable> fields should refer to the numeric soft and hard resource limit values. If the
+        second form is used, <replaceable>VALUE</replaceable> may specifiy a value that is used both as soft and hard
+        limit. In place of a numeric value the special string <literal>infinity</literal> may be used to turn off
+        resource limiting for the specific type of resource. This command line option may be used multiple times to
+        control limits on multiple limit types. If used multiple times for the same limit type, the last last use
+        wins. For details about resource limits see <citerefentry
+        project='man-pages'><refentrytitle>setrlimit</refentrytitle><manvolnum>2</manvolnum></citerefentry>. By default
+        resource limits for the container's init process (PID 1) are set to the same values the Linux kernel originally
+        passed to the host init system. Note that some resource limits are enforced on resources counted per user, in
```

Comment:
Hmm, what about DefaultLimitCPU= and other Default* settings? Should not specifying `--rlimit` mean that those limits are bumped above the setting for normal units?

Reply:
i'd really consider containers something seperate from the host, i.e. something that is more like KVM, and hence comes with defaults set up the same way at the kernel would do it on its own.

Note that this only resets the resource limits, which are generally per-user and per-process resources. Given that containers basically bring their own set of processes and users I think it makes a ton of sense to keep these resource limits specific to each container, and not make them something inherited from the host. 

I think conceptually rlimits are a very different beast from let's say cgroup resource limits, as those actually control resource use in a much more sensible and "integral" way, and are inherently something hierarchial. Those of course we should *not* reset, as it makes a ton of sense to apply limits to containers as a whole.

Or maybe there's a different way to see this: if you play around with DefaultLimitXYZ on the host, or set a resource limit for systemd-nspawn@.service or some instance thereof then this will have an effect on nspawn itself, but not on its payload, and I think that makes a ton of sense. The nspawn supervisor after all is part of the host, while the payload is a completely new environment, which the host should in almost all ways consider a single entity when it comes to resource management and most other forms of management. And given the rlimits are per-user/per-process limits they hence should not leak into this supposedly isolated environment by default.

And yes, this includes that containers might get higher rlimits than the user configured for the host but I think this is OK in the name of isolation

Reply:
Makes sense.